backend/cfg/vectorize.ml

```diff
@@ -178,61 +487,828 @@ end = struct
+                Some key)
+              else None
+            | Store _ ->
+              if Memory_accesses.all_adjacent memory_accesses instruction_ids
+              then (
+                Instruction.Id.Tbl.add computation_tree key node;
+                let built_dependencies = Array.make 1 (build_arg 0) in
+                update_node_dependencies built_dependencies)
+              else None
+            | Move | Const_int _ | Intop _ | Intop_imm _ | Specific _ ->
+              if Memory_accesses.inter_independent memory_accesses instructions
+              then (
+                Instruction.Id.Tbl.add computation_tree key node;
+                let built_dependencies =
+                  Array.mapi
+                    (fun arg_i _ -> build_arg arg_i)
+                    (Instruction.arguments last_instruction)
+                in
+                update_node_dependencies built_dependencies)
+              else None
+            | Alloc _ | Reinterpret_cast _ | Static_cast _ | Spill | Reload
+            | Const_float32 _ | Const_float _ | Const_symbol _ | Const_vec128 _
+            | Stackoffset _ | Intop_atomic _ | Floatop _ | Csel _
+            | Probe_is_enabled _ | Opaque | Begin_region | End_region
+            | Name_for_debugger _ | Dls_get | Poll ->
+              None))
+    in
+    let is_valid computation_tree =
+      (* Checks nodes can be grouped together, and instructions outside the tree
+         do not depend on the tree *)
+      let check cell =
+        match
+          DLL.value cell |> Instruction.id
+          |> Instruction.Id.Tbl.find_opt computation_tree
+        with
+        | None -> true
+        | Some (node : Node.t) ->
+          Memory_accesses.can_group ~remove:true memory_accesses body
+            node.instructions
+      in
+      let rec validate cell_option =
+        match cell_option with
+        | None -> true
+        | Some cell -> if check cell then validate (DLL.prev cell) else false
+      in
+      if DLL.last_cell body |> validate
+      then
+        let tree_is_not_dependency_of_the_rest_of_body =
+          let node_is_dependency_of (node : Node.t) =
+            List.fold_right
+              (fun instruction is_dependency_of ->
+                Dependency_graph.get_is_dependency_of dependency_graph
+                  instruction
+                |> Instruction.Id.Set.union is_dependency_of)
+              node.instructions Instruction.Id.Set.empty
+          in
+          let tree_is_dependency_of =
+            Instruction.Id.Tbl.fold
+              (fun _ node is_dependency_of ->
+                node_is_dependency_of node
+                |> Instruction.Id.Set.union is_dependency_of)
+              computation_tree Instruction.Id.Set.empty
+          in
+          DLL.to_list body |> List.map Instruction.id
+          |> Instruction.Id.Set.of_list
+          |> Instruction.Id.Set.inter tree_is_dependency_of
+          |> Instruction.Id.Set.is_empty
```

Comment:
(I wonder whether it would be
beneficial to avoid creating the
set.)

Reply:
I agree it is simpler to write this way,
but explicitly constructing the set
could be costly. It can be merged like
it is, but maybe live a CR-someday
about the possible cost (to be honest,
I don't know whether it will be an
issue in practice since your code is
working on individual blocks).

Reply:
done